src/node_embedding_api.h

```diff
@@ -0,0 +1,398 @@
+
+// Deletes the Node.js runtime instance.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_delete_runtime(node_embedding_runtime runtime);
+
+// Checks if the Node.js runtime is initialized.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_is_initialized(node_embedding_runtime runtime,
+                                      bool* result);
+
+// Sets the flags for the Node.js runtime initialization.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_set_flags(node_embedding_runtime runtime,
+                                 node_embedding_runtime_flags flags);
+
+// Sets the non-Node.js and Node.js CLI arguments for the Node.js runtime
+// initialization.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_set_args(node_embedding_runtime runtime,
+                                int32_t argc,
+                                const char* argv[],
+                                int32_t exec_argc,
+                                const char* exec_argv[]);
+
+// Sets the preload callback for the Node.js runtime initialization.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_on_preload(node_embedding_runtime runtime,
+                                  node_embedding_preload_callback preload_cb,
+                                  void* preload_cb_data);
+
+// Adds a new module to the Node.js runtime.
+// It is accessed as process._linkedBinding(module_name) in the main JS and in
+// the related worker threads.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_add_module(
+    node_embedding_runtime runtime,
+    const char* module_name,
+    node_embedding_initialize_module_callback init_module_cb,
+    void* init_module_cb_data,
+    int32_t module_node_api_version);
+
+// Initializes the Node.js runtime.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_initialize(node_embedding_runtime runtime,
+                                  const char* main_script);
+
+//------------------------------------------------------------------------------
+// Node.js runtime functions for the event loop.
+//------------------------------------------------------------------------------
+
+// Initializes the runtime to call the provided handler when the runtime event
+// loop has some work to do. It starts an observer thread that is stopped by the
+// `node_embedding_runtime_complete_event_loop` function call. This function
+// helps to integrate the Node.js runtime event loop with the host UI loop.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_on_event_loop_run_request(
+    node_embedding_runtime runtime,
+    node_embedding_event_loop_handler event_loop_handler,
+    void* event_loop_handler_data);
+
+// Runs the Node.js runtime event loop.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_runtime_run_event_loop(
```

Comment:
Even though libuv doesn't provide ABI guarantees, I think we should expose uv in embedder C API instead of wrapping it, allowing more flexibility of this C API since embedders have more control over `libnode`.

Reply:
While I agree with you that it can provide more flexibility, my concern is that non-C/C++ users have to bind to a much broader set of APIs which like you said may not be ABI safe.
Also, while we offer functions that seems to do the same as the `uv_run`, in practice they may be doing some more Node.js work such as draining the `v8::Isolate` work items.

Currently the `node_api.h` exposes the `uv_loop_t` associated with the `napi_env`. Technically, users can get it today. Though, we were discussing to deprecate it.

I also would like to explore a scenario where the UV loop is not used for processing the task queue, but we rather offer something like the V8 foreground task runner where the embedder API is responsible for pumping the task queue.

Thus, my proposal is to wait for a scenario where exposing the raw `uv_loop_t` is required before we add it.
Adding new API is easy, deprecating is more difficult.